Server-errors-and-solutions.md

```diff
@@ -0,0 +1,84 @@
+
+### Preventive action
+
+1. __Static analysis / compilation stage__: Having strong typing in the frontend will help catch some of these issues.
+2. __Manual testing__: Developers generally test only the "happy" paths in their feature. We should encourage them to test edge cases. Adoption of template users in testing mentioned [here](https://docs.google.com/document/d/1ktMagEKvUA6A0UmM_2CoENZPelqmjLmRjV7tWeo6w54/edit#heading=h.y3evobat23c5) should help uncover / test edge case scenarios.
+
+## Supporting old browsers
+
+Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: [#6909](https://github.com/oppia/oppia/pull/6909), [#6174](https://github.com/oppia/oppia/pull/6174), [#7926](https://github.com/oppia/oppia/pull/7926).
+
+### Preventive action
+
+1. __Static analysis / compilation stage__: Fix a JS standard that we expect all clients to support and specify a "target version" in typescript so that the code will be compiled to that version / JS standard.
+
+## Logic errors
+Logical errors in the code. Example: The "this" scope was referring to the global scope when it was supposed to refer to a local one ([#7873](https://github.com/oppia/oppia/pull/7873)). These are quite rare. Other examples of errors include: [#7872](https://github.com/oppia/oppia/issues/7872)(incorrect scope of "this"), [#7927](https://github.com/oppia/oppia/pull/7927).
+
+### Preventive action
+
+1. __Code review__: These issues need to be flagged during code reviews. They often happen when a developer is refactoring existing code but hasn’t fully tested their changes.
+
+## Data-related issue
+
+Issues that result in corrupt or missing data on the server. Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: [#6159](https://github.com/oppia/oppia/pull/6159), [#9632](https://github.com/oppia/oppia/pull/9632), [#7440](https://github.com/oppia/oppia/pull/7440).
+
+### Preventive action
+
+1. __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
+2. Testing with foreign / special characters (as mentioned under [Template Users](https://docs.google.com/document/d/1FYShgUHI0GM2I6eWe5B78PAQLvPxZxMFVjTivVoBrTg/edit#)).
```

Comment:
This doesn't match the [original doc](https://docs.google.com/document/d/1hIQIaFNVvyNiZBcCx0X7-_2lc0r1yvjxGfn4GJMjwEs/edit#heading=h.jlsplhp0gps7), and it's a little confusing. Performing destructive operations and using foreign/special characters are two types of manual testing that could help prevent this issue, so I think we actually want something like this:

> 1. __Manual testing__:
>    1. When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
>    2. Testing with foreign / special characters (as mentioned under [Template Users](https://docs.google.com/document/d/1FYShgUHI0GM2I6eWe5B78PAQLvPxZxMFVjTivVoBrTg/edit#)).

Reply:
done